### PR TITLE
move ctx.time assigment to the start of the serveConn loop

### PR DESCRIPTION
**Pull request by mmacdermaid**
@erikdubbelboer Any chance the time.Now() assignment can be moved to the start of the for loop, so the time involved with reading the request body and headers is included inside of it?

Setting it right before the handler is called doesn't do anyone good as they can just call it themselves once they have program flow.

### CODE REVIEW

**Comment by mmacdermaid** [comment]
I guess this more adequately does what I'm after.  By calling into client code after a response has been sent with the start time they can tag the finish time and send it off to any metrics collection they see fit.

I'm not 100% sure the call calling location is correct though, feel free to correct me if that wouldn't provide users with the correct timing.

**Comment by erikdubbelboer** [comment]
What you are doing here is possible with [`Server.ConnState`](https://godoc.org/github.com/valyala/fasthttp#Server.ConnState). If you add that hook [`StateIdle`](https://godoc.org/github.com/valyala/fasthttp#ConnState) indicates the response has been send and the connection went back to being idle. It's not far below where you call `RequestTimeHandler` in your pull request.

**Comment by erikdubbelboer** [comment]
I agree and I think the StateActive should be move to [here](https://github.com/valyala/fasthttp/blob/32940977fb67c66fa367b7fc8f59cfa0e4f80be7/server.go#L1987) so that it is triggered after reading only 4 bytes max. Not after reading the full headers and body.

Then it would be the same as the StateStart you proposed.

With that it should be possible to measure exactly how long each request takes with something like this:

```go
	starts := make(map[net.Conn]time.Time)
	var mu sync.Mutex

	s := fasthttp.Server{
		ConnState: func(conn net.Conn, state fasthttp.ConnState) {
			if state == fasthttp.StateActive {
				mu.Lock()
				starts[conn] = time.Now()
				mu.Unlock()
			} else if state == fasthttp.StateIdle || state == fasthttp.StateClosed {
				mu.Lock()
				took := time.Since(starts[conn])
				mu.Unlock()

				fmt.Printf("request took %v\n", took)
			}
		},
		// ...
	}
```
Would that work for you?

**Comment by mmacdermaid** [comment]
I now realize that doing this with connection state would require a mutex like you have posted above.  The idea I have in the PR avoids this by passing the attached time.  Adding another lock/unlock or some form of state tracking would not be ideal for this, as that means we're talking about doing it 2x on every request a connection sees which would be a lot of extra overhead.

I understand adding something like RequestTimeHandler starts to muddies of the waters of the sever object, because you could theoretically do this if we move the StateActive call (which I think is probably a good idea regardless).  

But I'm definitely alright with just forking master and adding it to an internal version if we can't come to a good solution here that works well for everyone.

File: server.go [hunk 1]
```diff
@@ -375,6 +375,10 @@ type Server struct {
 	// which will close it when needed.
 	KeepHijackedConns bool
 
+	// RequestTimeHandler will call into client code after a response has been sent with start time
+	// of the request.  This allows tracking of processing time for request.
+	RequestTimeHandler func(startTime time.Time)
```
[the remark this answers is not in the document]
Comment (mmacdermaid) [reply server.go:380]: > if you have several endpoints, and you want to monitor those separately

This is definitely also a drawback.

**Pull request closed without merging**